### PR TITLE
make public api consistent using each backend

### LINKED ISSUE #4033: The `wasmer` crate's `sys`, `js`, and `jsc` feature flags expose an inconsistent public API
### Describe the bug

As part of #4031 I use `wasmer::Artifact::is_deserializable()` to detect whether a compiled module has been LZW compressed or not. The `make test-js` build is currently failing because `Artifact` is only available with the `sys` target, not `js` and `jsc`.

```
error[E0433]: failed to resolve: could not find `Artifact` in `wasmer`
   --> lib/wasix/src/runtime/module_cache/filesystem.rs:159:16
    |
159 |     if wasmer::Artifact::is_deserializable(&leading_bytes) {
    |                ^^^^^^^^ could not find `Artifact` in `wasmer`
```

[(GitHub Actions log)](https://github.com/wasmerio/wasmer/actions/runs/5373019395/jobs/9746986067)

The root cause is that we aren't selective when re-exporting types from underlying implementations. We'll keep having these issues as long as we're doing `pub use sys::*`.

https://github.com/wasmerio/wasmer/blob/77898a7767eb7057834fdc1a80b239e6b68cf44e/lib/api/src/lib.rs#L448-L464

I'm guessing this regression was originally introduced in #3556 and we didn't notice it when JSC support was added in #3825.

### Proposed Solution

The best way to fix this would be to change the way we re-export things to always re-export the same types. That way backend-specific items like `Artifact` won't be part of our universal `wasmer` API.

```rust
cfg_if::cfg_if! {
  if #[cfg(feature = "sys")] {
    pub mod sys;
    use sys as imp;
  } else if #[cfg(feature = "js")] {
    pub mod js;
    use js as imp;
  } else if #[cfg(feature = "jsc")] {
    pub mod jsc;
    use jsc as imp;
  } else {
    // Use a fallback implementation
     #[path = "unsupported/mod.rs"]
    mod imp;
  }
}

pub use imp::{Module, Instance, ...};
```

This has the added benefit that users will still be able to access items from the `sys`, `js`, and `jsc` modules as an escape hatch. We'll also get those nice *"Available on crate feature sys only"* annotations from rustdoc if we set `#![cfg_attr(docsrs, feature(doc_cfg))]`.

Given we literally *just* released Wasmer 4.0 with a bunch of breaking changes and this will remove items from `wasmer`'s public API, I don't think this is feasible.

### Alternative Solution 

Another solution is to create a dummy implementation of `wasmer::Artifact` (and all the other missing types) that is used whenever the `js` and `jsc` feature flags are active. 

This is similar to how the Rust standard library has a `std::sys` module containing all OS-specific details, and whenever a particular target doesn't implement something it'll fall back to the implementation in `std::sys::unsupported`.

### Alternative Solution 2

A quickfix could be to mark the re-export of `Artifact` as deprecated, saying people should use the `wasmer_compiler` crate directly.

https://github.com/wasmerio/wasmer/blob/77898a7767eb7057834fdc1a80b239e6b68cf44e/lib/api/src/sys/mod.rs#L17

This approach kinda works, but it'll become a constant game of wack-a-mole as long as those `pub use sys::*` exports exist.

### PR DESCRIPTION
**Pull request by maminrayej**
This PR tries to combine all the suggestions in #4033:

- By hiding `sys`, `js` and `jsc` behind the `backend` module, it makes sure all three backends expose the same API.
- In order to not break the public API, stub structs, traits, ... are provided via the `stub` module. `js` and `jsc` use this module to expose the same API as `sys` by re-exporting items that they don't have an implementation for.
- Unfortunately it seems using `#[deprecated]` with re-exports doesn't work in Rust yet and is a known [issue](https://github.com/rust-lang/rust/issues/30827). So I didn't have a way to incorporate this suggestion from the original issue.

Resolves #4033.

### CODE REVIEW

**Comment by syrusakbary** [comment]
> By hiding sys, js and jsc behind the backend module, it makes sure all three backends expose the same API.

I don't see how. This PR just repositioning the files, creating a `lib/api/src/backend/unsupported.rs` (should those  types really exist in other platforms?).

**Comment by maminrayej** [comment]
> I don't see how. This PR just repositioning the files, creating a lib/api/src/backend/unsupported.rs (should those types really exist in other platforms?).

The problem at hand are the items that are exposed as part of `lib/api` that depend on a specific backend like `sys`. One example is what Michael encountered. For example `Artifact` is exposed when `sys` is enabled so one can use it and then expect it to work, just to find out their project won't build with `js` or `jsc`. This is the inconsistency that this PR is trying to solve. From here on out we have two options:

1. Remove things like `Artifact` from our api and put it behind a feature gated module like `sys`. This will make the api exposed by `sys` consistent with `js` and `jsc`, but requires a breaking change. I don't think we can afford it right now.
2. Add a temporary stub implementation of `Artifact` for `js` and `jsc`. This also makes the api exposed by these backends consistent with each other. It doesn't require a breaking change and is the way Rust std manages OS specific things for example.

This PR tries to implement the second solution. Also, in order to prevent this mistake from happening again in the future, it explicitly re-exports everything from the backend. For example, assume we add `Foo` and expose it via `sys`. We have to expose it via `backend` too. But, since `backend` expects all of its `imps` to expose `Foo`, `lib/api` won't get compiled until we provide an impl of `Foo` for all backends, preventing any inconsistency to happen again.

So in conclusion, this PR makes the api exposed by different backends consistent without a breaking change, and prevents inconsistencies in the future.

Note that this PR is marked as draft and is still completely ready to be reviewed.

File: lib/api/src/backend/unsupported.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#![allow(dead_code)]
+
+use std::marker::PhantomData;
```
Comment (syrusakbary) [inline lib/api/src/backend/unsupported.rs:3]: We chatted about not doing this strategy (as it can infinitely open up the API surface), but only exposing the sys API through the main lib as deprecated exported fields to prompt the user to use `::sys` instead

File: lib/api/src/lib.rs [hunk 2]
```diff
@@ -446,11 +446,36 @@ mod value;
 pub mod vm;
 
 #[cfg(feature = "sys")]
-mod sys;
+/// sys
+pub mod sys;
 
 #[cfg(feature = "sys")]
 pub use sys::*;
```
Comment (syrusakbary) [inline lib/api/src/lib.rs:453]: Do we want this?

File: lib/api/src/lib.rs [hunk 3]
```diff
@@ -446,11 +446,36 @@ mod value;
...
 
+#[cfg(feature = "sys")]
+#[deprecated(note = "Use sys::Artifact instead")]
```
Comment (syrusakbary) [inline lib/api/src/lib.rs:456]:
```suggestion
#[deprecated(note = "wasmer::Artifact is deprecated, use wasmer::sys::Artifact instead")]
```

File: lib/api/src/lib.rs (hunk 2 above)
Comment (maminrayej) [reply lib/api/src/lib.rs:453]: Well we're saying `use wasmer::sys::Artifact instead`, so this module should be available for users to be able to migrate.

**Review by syrusakbary — APPROVED** [review]
